_posts/2017-10-11-cross-site-request-forgery.markdown

```diff
@@ -0,0 +1,217 @@
+
+What all of this means is that if the attacker directs an end-user to *evil.com*, they will not be able to make AJAX requests on behalf of the user to *example.com* under normal circumstances. Given that the same-origin policy can be found too restrictive at times, there is a separate feature which allows exceptions from it, called [CORS] (Cross Origin Resource Sharing). The CORS standard defines a set of HTTP headers that can be set by applications to allow AJAX requests from different origins.
+
+Let us assume the worst possible case where the application in question sets the following CORS headers on all of its pages:
+
+```html
+Access-Control-Allow-Origin: http://evil.com
+Access-Control-Allow-Credentials: true
+```
+
+This means that the application's pages are accessible via AJAX from the *evil.com* domain, and the credentials are also made available during the AJAX requests. In this scenario the attacker could embed the following JavaScript code into a page on *evil.com*:
+
+```javascript
+$(document).ready(function() {
+    $.ajaxSetup({crossDomain: true, xhrFields: {withCredentials: true}});
+    $.get('http://example.com/update').done(function(data) {
+        var csrf = data.match(/.*name="csrf"[^>]*value="([^"]*)".*/)[1];
+        $.ajax({
+            type: "POST",
+            url: "http://example.com/update",
+            data: "username=demo&password=pwned&csrf=" + csrf
+        }).done(function(data) {
+            console.log("Hack successful");
+        });
+    });
+});
+```
+
+The gist of this exploit is that it first sends a request as the end-user to the update page to obtain the CSRF token from the HTML response. Once this is complete, there is a second request made that includes the CSRF token and the other input fields to actually execute the operation. A couple of things I should highlight about this attack:
+
+* I was only able to perform the operation, because of the **Access-Control-Allow-Credentials** header. Without it, the AJAX call would not have included the end-user's cookies, and as such the call would have been unauthenticated.
+* Fortunately the CORS spec is very sensible when it comes to **Access-Control-Allow-Credentials**, it cannot be used along with **Access-Control-Allow-Origin: \***, this should prevent misconfigurations exposing applications to these sort of attacks.
+* This kind of attack would just as well work if script can be injected into the same origin as the application (and then you wouldn't need CORS headers either).
```

Comment:
Maybe worth noting that all protections against CSRF are defeated if XSS is possible.